### PR TITLE
Updating module to allow for tmt import of 'metadata' file for #1002

### PR DESCRIPTION
**Pull request by pfdaly**
Module has been updated to allow for tmt to import 'metadata' file as well as 'Makefile'. If both exist in the directory, priority is given to 'metadata' file.

The command remains the same, 'tmt test import --makefile' to allow for backwards compatibility.

# Existing code behaviour.
# Makefile, Purpose file, and main.fmf present in the directory.

[root@pdaly parent]# tmt test import --makefile
Checking the '/root/auto-kernel/tmt_fork/tmt/tests/test/import/data/parent' directory.
Makefile found in '/root/auto-kernel/tmt_fork/tmt/tests/test/import/data/parent/Makefile'.
task: /tmt/smoke
summary: Simple smoke test
test: ./runtest.sh
framework: shell
contact: James Bond <007@MI6.com>
component: tmt
duration: 5m
environment:
{'AVC_ERROR': '+no_avc_check',
 'CONTEXT': 'distro=fedora',
 'TEST': 'one two three'}
require: tmt
recommend: fmf
tag: multihost
relates: https://bugzilla.redhat.com/show_bug.cgi?id=1234567
relates: https://bugzilla.redhat.com/show_bug.cgi?id=9876543
Purpose found in '/root/auto-kernel/tmt_fork/tmt/tests/test/import/data/parent/PURPOSE'.
description:
Just run 'tmt --help' to make sure the binary is sane.
This is really that simple. Nothing more here. Really.
Nitrate Major (851968): Unspecified GSS failure.  Minor code may provide more information, Minor (2529638944): Ticket expired
[root@pdaly parent]#

# Implemented code behaviour
# Makefile, Purpose file, and main.fmf present in the directory.

[root@pdaly parent]# tmt test import --makefile
Checking the '/root/auto-kernel/tmt_fork/tmt/tests/test/import/data/parent' directory.
Makefile found in '/root/auto-kernel/tmt_fork/tmt/tests/test/import/data/parent/Makefile'.
task: /tmt/smoke
summary: Simple smoke test
test: ./runtest.sh
framework: shell
contact: James Bond <007@MI6.com>
component: tmt
duration: 5m
environment:
{'AVC_ERROR': '+no_avc_check',
 'CONTEXT': 'distro=fedora',
 'TEST': 'one two three'}
require: tmt
recommend: fmf
tag: multihost
relates: https://bugzilla.redhat.com/show_bug.cgi?id=1234567
relates: https://bugzilla.redhat.com/show_bug.cgi?id=9876543
relates: https://bugzilla.redhat.com/show_bug.cgi?id=2222222
Purpose found in '/root/auto-kernel/tmt_fork/tmt/tests/test/import/data/parent/PURPOSE'.
description:
Just run 'tmt --help' to make sure the binary is sane.
This is really that simple. Nothing more here. Really.
Nitrate Major (851968): Unspecified GSS failure.  Minor code may provide more information, Minor (2529638944): Ticket expired
[root@pdaly parent]#

# metadata, Purpose file, and main.fmf present in the directory.

[root@pdaly parent]# tmt test import --makefile
Checking the '/root/auto-kernel/tmt_fork/tmt/tests/test/import/data/parent' directory.
metadata found in '/root/auto-kernel/tmt_fork/tmt/tests/test/import/data/parent/metadata'.
task: /tmt/smoke
summary: Simple smoke test
test: ./runtest.sh
framework: shell
contact: Napoleon Solo <TheManFrom@UNCLE.com>
component: tmt
duration: 5m
recommend: fmf tmt
Purpose found in '/root/auto-kernel/tmt_fork/tmt/tests/test/import/data/parent/PURPOSE'.
description:
Just run 'tmt --help' to make sure the binary is sane.
This is really that simple. Nothing more here. Really.
Nitrate Major (851968): Unspecified GSS failure.  Minor code may provide more information, Minor (2529638944): Ticket expired
[root@pdaly parent]#

### CODE REVIEW

**Comment by psss** [comment]
Thank you for looking into this! Would you mind adding at least a basic test coverage for the new feature? It would help with the review to have example data for experimenting. You could extend the existing test [/tests/test/import](https://github.com/psss/tmt/tree/master/tests/test/import) or write a new test.

Regarding the options, currently we have an enable/disable option for each metadata source:

```
  --nitrate / --no-nitrate    Import test metadata from Nitrate
  --purpose / --no-purpose    Migrate description from PURPOSE file
  --makefile / --no-makefile  Convert Beaker Makefile metadata
```

I think it would be good to introduce similar options for the `restraint` metadata parsing as well. @lukaszachy, @thrix, what do you think?

**Comment by lukaszachy** [comment]
Should be `--restraint / --no-restraint  Convert restraint metadata file`

Now what should be the default and how fatal would be missing piece ... 
IMO from user PoV `tmt test import .` should just work. However If there is explicit --restraint on the commandline I'd expect failure if 'metadata' file is missing. But that could be part of another PR.

**Review by lukaszachy — CHANGES_REQUESTED** [review]
Hi @pfdaly, thanks for the PR. I believe the logic for `repoRequires` is wrong and `softDependencies` are missing.
Makefile maps Requires -> recommend, RhtsRequire -> require
and metadata should do repoRequires -> require, softDependencies -> recommend

So that makes `regex_requires` technically a recommend/softDependencies.
IMO the `regex_XXX` names should called after target destination - so singular form.
This way later one will use regex_require to fill data['require'] and regex_recommend to fill data['recommend']

**Comment by pfdaly** [comment]
Hi @lukaszachy,
My latest push fixes the variable names of the regex patterns to match the key names of the data dictionary.

The change also includes the mapping of the metadata repoRequires > require & softDependencies -> recommend.

**Review by lukaszachy — CHANGES_REQUESTED** [review]
Thank you, few comments which I didn't see before.
Also test.sh is mixing tabs and spaces -> rest of the file is with spaces so please retab ;)

Hopefully nitpick - commit header is used in release notes so please set to something shorter https://tmt.readthedocs.io/en/stable/contribute.html#commits

IMO among the lines of "Allow import from restraint's metadata file"

**Comment by lukaszachy** [comment]
Also please do rebase to the latest main. It might fix the tests.

File: tests/test/import/test.sh [hunk 1]
```diff
@@ -34,6 +35,65 @@ rlJournalStart
...
+
+
+    rlPhaseStartTest 'Verify error returned when no Makefile exists.'
+        rlRun "mv $tmp/data/parent/child/Makefile $tmp/data/parent/child/.Makefile" 0 "Removing Makefile in tmp directory"
+	rlRun 'tmt test import --no-nitrate 2>&1 | tee output' 2
+        rlAssertGrep 'Unable to find Makefile' 'output'
+    rlPhaseEnd
+
+    rlPhaseStartTest 'Verify error returned when no Restraint metadata file exists.'
+        rlRun "mv $tmp/data/parent/child/metadata $tmp/data/parent/child/.metadata" 0 "Removing restraint metadata file in tmp directory"
```
Comment (lukaszachy) [inline tests/test/import/test.sh:90]: There is rlFileBackup & rlFileRestore in beakerlib. So you might call `rlFileBackp <path>`, then `rm <path>` and in the end just call `rlFileRestore` to restore all backuped files.

File: tmt/convert.py [hunk 2]
```diff
@@ -281,13 +241,20 @@ def read(path, makefile, nitrate, purpose, disabled, types):
...
             pass
-        # Duration
-        try:
-            data['duration'] = re.search(
-                r'^TestTime:\s*(.*)', testinfo, re.M).group(1)
-            echo(style('duration: ', fg='green') + data['duration'])
-        except AttributeError:
-            pass
+    else:
+        data['component'] = 'tmt'
```
[the remark this answers is not in the document]
Comment (lukaszachy) [reply tmt/convert.py:245]: Please do not set this value if it cannot be determined properly. Once can set it in top main.fmf to have it available in all children nodes. Value 'tmt' is not a correct one.

**Review by lukaszachy — APPROVED** [review]
Thank you very much for the PR, I'll rebase and merge.

**Comment by lukaszachy** [comment]
So, hopefully I fixed pre-commit format issue and tests now pass

**Comment by pfdaly** [comment]
Great, thanks for that @lukaszachy.
It's odd that the pre-commit failed with a pep8 error as I'd executed Pycodestyle on it with a pass result prior to pushing. There must be some detection differences between the two.

**Comment by lukaszachy** [comment]
@pfdaly Maybe - I found out that even different versions of checker might provide different opinions about code format.
Best is to set up  [pre-commit](https://tmt.readthedocs.io/en/stable/contribute.html#develop) which should be the same locally and in PR

**Comment by BeeGrech** [comment]
Please forgive my question if covered - I dont see where metadata dependencies are processed.

**Comment by lukaszachy** [comment]
Ah, now I see it in https://restraint.readthedocs.io/en/latest/tasks.html#dependencies
So `dependencies` are addition to `repoRequires` ?

**Comment by BeeGrech** [comment]
repoRequires is actually like a preparation script-
[](https://restraint.readthedocs.io/en/latest/tasks.html?highlight=reporequires#reporequires)repoRequires
A semicolon-delimited (;) list of additional tasks needed for this task to run.

now another question, I dont think tests support prepare steps...

**Pull request merged**